dashboard/config/routes.rb

```diff
@@ -323,6 +323,7 @@
       get :search
     end
   end
+  get 'programming_expressions/:environment/:expression', to: 'programming_expressions#show'
```

Comment:
I have a few gripes here:

1. it seems like it would be clearer to have this route use a different controller action, and extract common functionality as needed. this one could become show_by_name, or the other could change to show_by_id
2. it would be more railsy to defines this using a `member` block inside the `resources` block
3. it seems a bit fragile that GETs for `programming_expressions/:id` and `programming_expressions/:environment/:expression` are disambiguated only by the number of parameters, especially because adding another `member` operation named `get 'foo'` could create a route like `programming_expressions/:id/foo`, which could(?) conflict with this new route

putting these all together, I would recommend something like this:

```
resources :programming_expressions ...
  member do
    get 'show_by_name/:environment/:expression', action: 'show_by_name'
```
which will hit the `show_by_name` method in the controller. I don't see an exact example of this in our routes file, so please feel free to ping me if I am off on the syntax.

Alternately... I'd be open to a discussion on whether it's worth spending energy on trying to follow all the rails norms and conventions. As with most rails things I find that I like them after a startup cost of learning what the conventions are supposed to be.

Reply:
on point 3, it looks like the new route may already conflict with GET 'programming_expressions/1/edit'. also for background here is the guide I'm looking at for rails routes: https://guides.rubyonrails.org/routing.html

Reply:
Thanks for the feedback! I'll change the route to be something a little less ambiguous :)

Thanks for the tip on using a `member` block -- I had a bit of trouble wrapping my head around how to use all of the syntax here.

Reply:
oh, `member` might not be right because that will try 'programming_expression/1/.../...'. `collection` might be what you want instead. I'd have to brush up on the rails routing guide to be sure.

Reply:
haha yeah that's what I think I ran into when I tried this briefly and what I ran into just now. Playing with it now, but it may end up in `collection`

Reply:
Okay, since `collection` felt a bit weird to me, I changed this to `programming_environments/:programming_environment_name/programming_expressions/:programming_expression_name`. It's a bit of a mouthful but follows the routing patterns we use for script/lessons/levels and should work for now!